docs/sdk.md

```diff
@@ -0,0 +1,88 @@
+Calypso SDK
+===========
+
+Calypso SDK is an early stage tool that's goal is to build, visualize, test, and deliver extensions from a single source.
+
+## Using SDK CLI
+
+```bash
+npm run sdk -- --help
+```
+
+Note: It's also possible to run the SDK command "globally" by linking within the Calypso repository with [`npm link`](https://docs.npmjs.com/cli/link). After running this command you can replace all invocations of `npm run sdk --` in the examples below with `calypso-sdk` and may do so from any other directory in the filesystem:
+```
+calypso-sdk --help
+```
+
+## SDK modules
+
+SDK can be extended with modules to perform different tasks. Currently we have only one task; build Gutenberg extensions.
+
+### Gutenberg extensions
+
+SDK module to build [Gutenberg](https://wordpress.org/gutenberg/handbook/) extensions.
+
+See usage instructions:
+
+```
+npm run sdk -- gutenberg --help
+```
+or
+```
+npm run sdk:gutenberg -- --help
+```
+
+## Extending the SDK
+
+CLI API definitions live in `bin/sdk-cli.js` and are constructed using [Yargs](http://yargs.js.org/) library.
+
+Add new modules to SDK by extending the CLI using [`yargs.command`](http://yargs.js.org/docs/#api-commandcmd-desc-builder-handler).
+
+Here is a very basic example to follow:
+
+In `bin/sdk-cli.js`:
+```js
+const example = require('./sdk/example.js');
+
+yargs.command( {
+	command: 'example',
+	desc: 'Build example extension',
+	builder: yargs =>
+		yargs.options( {
+			'output-dir': {
+				description:
+					'Output directory for the built assets. Intermediate directories are created as required.',
+				type: 'string',
+				coerce: path.resolve,
+				requiresArg: true,
+			},
+		} ),
+	handler: argv => build( example, argv ),
+} );
+```
+
+`bin/sdk/example.js`:
+```js
+const path = require( 'path' );
+
+exports.config = ( {
+	{ argv: { outputDir },
+	getBaseConfig,
+} ) => {
+	const baseConfig = getBaseConfig();
+	return {
+		...baseConfig,
+		entry: path.resolve( __dirname, '..', 'client', 'example' ),
+		output: {
+			path: outputDir,
+			filename: 'example-build.js',
+		},
+	};
+};
+```
```

Comment:
can we add a section here like "when should I add a new target/project" and then make the answer something like "don't" 😄 

I'm aware of at least one or two additional build targets we need - @ockham and I spoke about a generic "npm package" target - but I want to communicate here that this isn't the place to create a new config for every project but rather this is a place to create configurations for entirely separate classes of projects

the notifications panel may seem like a counter-example and maybe it is - I don't know. it's qualitatively different because it relies on building an HTML file as well. theoretically we could iterate on that idea to generalize as the need arises.

Reply:
Yeah, might be we can eventually stop dividing the SDK in "targets" and simply have generic tasks like "build" or "test".

I added this:

> Adding new build targets for the SDK is straightforward but should be done with some consideration. Keeping the SDK as generic as possible is essential, and thus this isn't a place to create a new config for every project. Preferably this is a place to create configurations for entirely separate classes of projects.

What do you think?

Reply:
> stop dividing the SDK in "targets" and simply have generic tasks like "build" or "test"

I doubt that because the inherent purpose is to produce semantics for each kind of output. unless we decide to build a monolith SDK where every project is rebuilt on every commit, which I think is extremely unlikely.

> What do you think?

that seems good to me. truth be told we will probably expect a small proliferation of targets commensurate with the number of high-level products we have at Automattic. one for WooCommerce, one for Jetpack, one for the array of WordPress plugins we produce maybe, one for the array of `npm` packages we publish

For now I think the bit about `this isn't a place to create a new config for every project` should be clear enough. Maybe we should make it bold just to make it stick out more though.

Reply:
>we will probably expect a small proliferation of targets commensurate with the number of high-level products we have at Automattic. one for WooCommerce, one for Jetpack, one for the array of WordPress plugins we produce maybe, one for the array of npm packages we publish

That's probably fine for the starters in a way that it'll help everyone to see where they overlap and move as much of the overlapping logic to `sdk-cli.js` instead of keeping it under `./sdk/*`